File: chrome/browser/offline_items_collection/offline_content_aggregator_factory.cc
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/offline_items_collection/offline_content_aggregator_factory.h"

#include "base/memory/singleton.h"
#include "chrome/browser/profiles/incognito_helpers.h"
#include "components/keyed_service/content/browser_context_dependency_manager.h"
#include "components/offline_items_collection/core/offline_content_aggregator.h"
#include "content/public/browser/browser_context.h"

// static
OfflineContentAggregatorFactory*
OfflineContentAggregatorFactory::GetInstance() {
  return base::Singleton<OfflineContentAggregatorFactory>::get();
}

// static
offline_items_collection::OfflineContentAggregator*
OfflineContentAggregatorFactory::GetForBrowserContext(
    content::BrowserContext* context) {
  DCHECK(!context->IsOffTheRecord());
  return static_cast<offline_items_collection::OfflineContentAggregator*>(
      GetInstance()->GetServiceForBrowserContext(context, true));
}

OfflineContentAggregatorFactory::OfflineContentAggregatorFactory()
    : BrowserContextKeyedServiceFactory(
          "offline_items_collection::OfflineContentAggregator",
          BrowserContextDependencyManager::GetInstance()) {}

OfflineContentAggregatorFactory::~OfflineContentAggregatorFactory() = default;

KeyedService* OfflineContentAggregatorFactory::BuildServiceInstanceFor(
    content::BrowserContext* context) const {
  return new offline_items_collection::OfflineContentAggregator();
}

content::BrowserContext*
OfflineContentAggregatorFactory::GetBrowserContextToUse(
    content::BrowserContext* context) const {
  return chrome::GetBrowserContextRedirectedInIncognito(context);
}

File: chrome/android/java/src/org/chromium/chrome/browser/download/items/OfflineContentAggregatorFactory.java
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.download.items;

import org.chromium.chrome.browser.profiles.Profile;
import org.chromium.components.offline_items_collection.OfflineContentProvider;

/**
 * Basic factory that creates and returns an {@link OfflineContentProvider} that is attached
 * natively to {@link Profile}.
 */
public class OfflineContentAggregatorFactory {
    private OfflineContentAggregatorFactory() {}

    /**
     * Used to get access to the {@link OfflineContentProvider} associated with {@code profile}.
     * The same {@link OfflineContentProvider} will be returned for the same {@link Profile}.
     * @param profile The {@link Profile} that owns the {@link OfflineContentProvider}.
     * @return An {@link OfflineContentProvider} instance.
     */
    public static OfflineContentProvider forProfile(Profile profile) {
        return nativeGetOfflineContentAggregatorForProfile(profile);
    }

    private static native OfflineContentProvider nativeGetOfflineContentAggregatorForProfile(
            Profile profile);
}